sdk/dotnet/Pulumi.Tests/Deployment/DeploymentRunnerTests.cs

```diff
@@ -0,0 +1,209 @@
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+using Xunit;
+
+using Pulumi.Testing;
+using Pulumi.Tests.Mocks;
+
+namespace Pulumi.Tests
+{
+    public class DeploymentRunnerTests
+    {
+        [Fact]
+        public async Task WorksUnderStress()
+        {
+            var resources = await Deployment.TestAsync<StressRunnerStack>(new EmptyMocks());
+            var stack = (StressRunnerStack)resources[0];
+            var result = await stack.RunnerResult;
+            Assert.Equal(0, result);
+        }
+
+        class StressRunnerStack : Stack
+        {
+            public Task<int> RunnerResult { get; private set; }
+            public StressRunnerStack()
+            {
+                var runner = Pulumi.Deployment.Instance.Internal.Runner;
+
+                for (var i = 0; i < 100; i++)
+                {
+                    runner.RegisterTask($"task{i}", Task.Delay(100 + i));
+                }
+
+                this.RunnerResult = runner.RunAsync<EmptyStack>();
+            }
+        }
+
+        [Fact]
+        public async Task TerminatesEarlyOnException()
+        {
+            var deployResult = await Deployment.TryTestAsync<TerminatesEarlyOnExceptionStack>(new EmptyMocks());
+            Assert.NotNull(deployResult.Exception);
+            if (deployResult.Exception! is AggregateException)
+            {
+                // The "Deliberate test error" should really be thrown
+                // only once, but unfortunately observing the multiple
+                // exceptions case in CI, so adding this clause to
+                // ensure all of them are test errors, and we just
+                // have unintentional duplication.
+                var e = (AggregateException)deployResult.Exception!;
+                foreach (var err in e.InnerExceptions)
+                {
+                    Assert.IsType<RunException>(err);
+                    Assert.Contains("Deliberate test error", err.Message);
+                }
+            }
+            else
+            {
+                Assert.IsType<RunException>(deployResult.Exception!);
+                Assert.Contains("Deliberate test error", deployResult.Exception!.Message);
+            }
+
+            var stack = (TerminatesEarlyOnExceptionStack)deployResult.Resources[0];
+            var result = await stack.RunnerResult;
+            Assert.Equal(0, result);
+        }
+
+        class TerminatesEarlyOnExceptionStack : Stack
+        {
+            public Task<int> RunnerResult { get; private set; }
+
+            public TerminatesEarlyOnExceptionStack()
+            {
+                var runner = Pulumi.Deployment.Instance.Internal.Runner;
+                var error = new Exception("Deliberate test error");
+
+                Task task1 = Task.FromException(error);
+                Task task2 = Task.Delay(100);
+
+                runner.RegisterTask("task1", task1);
```

Comment:
Minor thing but I would swap the order of registrations to "prove" that the exception is returned even if it's registered later.